File: Day20-25/24.py
import functools
from multiprocessing import Process, Queue


def parse_component(line):
    a, b = line.split('/')
    return (int(a), int(b))


@functools.lru_cache(None)
def _strongest_chain(components, next_comp):
    max_strength = 0
    for c in list(components):
        if next_comp in c:
            c_next_comp = c[0] if c[1] == next_comp else c[1]
            s = _strongest_chain(components - frozenset([c]), c_next_comp)
            if s + next_comp > max_strength:
                max_strength = s + next_comp
    return max_strength + next_comp


def find_strongest_bridge(components, q):
    max_strength = 0

    for c in list(components):
        if 0 in c:
            non_zero = c[0] if c[1] == 0 else c[1]
            s = _strongest_chain(components - frozenset([c]), non_zero)
            if s > max_strength:
                max_strength = s
    q.put(max_strength)


@functools.lru_cache(None)
def _longest_chain(components, next_comp):
    max_strength, max_length = 0, 0
    for c in list(components):
        if next_comp in c:
            c_next_comp = c[0] if c[1] == next_comp else c[1]
            s, l = _longest_chain(components - frozenset([c]), c_next_comp)
            if l > max_length or (l == max_length and s + next_comp > max_strength):
                max_strength = s + next_comp
                max_length = l
    return max_strength + next_comp, max_length + 1


def find_longest_bridge(components, q):
    max_strength, max_length = 0, 0

    for c in list(components):
        if 0 in c:
            non_zero = c[0] if c[1] == 0 else c[1]
            s, l = _longest_chain(components - frozenset([c]), non_zero)
            if l > max_length or (l == max_length and s > max_strength):
                max_strength = s
                max_length = l
    q.put(max_strength)


if __name__ == '__main__':
    with open('24.txt') as f:
        lines = f.readlines()
    components = frozenset(map(parse_component, lines))

    p1_q, p2_q = Queue(), Queue()

    p1_proc = Process(target=find_strongest_bridge, args=(components, p1_q))
    p2_proc = Process(target=find_longest_bridge, args=(components, p2_q))
    p1_proc.start()
    p2_proc.start()

    print("Part 1: {}".format(p1_q.get()))
    print("Part 2: {}".format(p2_q.get()))

    p1_proc.join()
    p2_proc.join()

File: Day20-25/24.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type Component struct {
	pins [2]int
	used bool
}

func parseComponents(lines []string) map[int][]*Component {
	c := make(map[int][]*Component, len(lines))

	for _, line := range lines {
		nums := strings.Split(line, "/")
		a, _ := strconv.Atoi(nums[0])
		b, _ := strconv.Atoi(nums[1])

		comp := Component{
			pins: [2]int{a, b},
			used: false,
		}

		arr, ok := c[a]
		if !ok {
			c[a] = []*Component{&comp}
		} else {
			c[a] = append(arr, &comp)
		}

		arr, ok = c[b]
		if !ok {
			c[b] = []*Component{&comp}
		} else {
			c[b] = append(arr, &comp)
		}
	}

	return c
}

func strongestBridgeIter(components *map[int][]*Component, nextComp int) int {
	maxStrength := 0
	for i, comp := range (*components)[nextComp] {
		if comp.used {
			continue
		}

		var currNextComp int
		if comp.pins[0] == nextComp {
			currNextComp = comp.pins[1]
		} else {
			currNextComp = comp.pins[0]
		}

		(*components)[nextComp][i].used = true
		strength := strongestBridgeIter(components, currNextComp)
		(*components)[nextComp][i].used = false

		if strength+nextComp > maxStrength {
			maxStrength = strength + nextComp
		}
	}
	return maxStrength + nextComp
}

func findStrongestBridge(components map[int][]*Component, c chan int) {
	c <- strongestBridgeIter(&components, 0)
}

func longestBridgeIter(components *map[int][]*Component, nextComp int) (int, int) {
	maxStrength := 0
	maxLength := 0
	for i, comp := range (*components)[nextComp] {
		if comp.used {
			continue
		}

		var currNextComp int
		if comp.pins[0] == nextComp {
			currNextComp = comp.pins[1]
		} else {
			currNextComp = comp.pins[0]
		}

		(*components)[nextComp][i].used = true
		strength, length := longestBridgeIter(components, currNextComp)
		(*components)[nextComp][i].used = false

		if length > maxLength || (length == maxLength && strength+nextComp > maxStrength) {
			maxStrength = strength + nextComp
			maxLength = length
		}
	}
	return maxStrength + nextComp, maxLength + 1
}

func findLongestBridge(components map[int][]*Component, c chan int) {
	strength, _ := longestBridgeIter(&components, 0)
	c <- strength
}

func main() {
	dat, _ := ioutil.ReadFile("24.txt")
	lines := strings.Split(string(dat), "\n")

	c1 := make(chan int)
	c2 := make(chan int)

	go findStrongestBridge(parseComponents(lines), c1)
	go findLongestBridge(parseComponents(lines), c2)

	fmt.Printf("Part 1: %d\n", <-c1)
	fmt.Printf("Part 2: %d\n", <-c2)
}
